pallets/bridge_transfer/src/lib.rs

```diff
@@ -130,7 +137,19 @@ pub mod pallet {
 			} else {
 				min_fee
 			};
-			T::Currency::transfer(&source, &bridge_id, amount + fee, AllowDeath)?;
+			let imbalance = T::Currency::withdraw(
```

Comment:
This is not atomic. Consider the situation where the sender doesn't have enough token in their account, it will first deduct the fee from the sender. Assuming it succeeded, then in the next step, the transfer can fail. Although the tx failed, the fee is still deducted.

If you add a unit test with `ensure_noop!(...)`, it will report as an error, because usually it's assumed that a failed Substrate call shouldn't modify the state on the blockchain.

A solution could be, we first calculate the total imbalance (amount + fee), and call `withdraw` to get it from the sender's account. Then we split the amount into the fee, and the transfer balance, then let the Treasury and the bridge pool to absorb the imbalance.

Also, a unit test is required to make sure it works as expected.

Reply:
It's true that there is not atomic, but instead of withdraw  (amount + fee) in advance, we have a better choice is to use [ensure_can_withdraw](https://github.com/paritytech/substrate/blob/6d0c04d58fc614f61072f6dde1a1716fb8072a44/frame/balances/src/lib.rs#L1435)  provided by Currency trait. It a dry run, just  gurantee there has enough free balance in user account.

Reply:
Yes, that's another option

Reply:
Finally i jsut use `ensure!(free_balance >= (amount + fee), Error::<T>::InsufficientBalance);` to guarantee the balance in account